### ISSUE TITLE
Frequent pop-up on save: "Getting code actions: Go, Go, ..."

### ISSUE THREAD

**Issue by stamblerre** [golang/vscode-go#236]
**Copied from https://github.com/microsoft/vscode-go/issues/3179:**

If you encounter this issue, please try the following steps to troubleshoot:

1. Check your other installed extensions and see if disabling them but the `Go` extension helps.
 **Go Group Imports**, **Spell Right**, and **Go Test Explorer** have both been identified as causing slow saves.
2. Your workspace may have errors which are causing gopls to be slow. Capture your gopls logs ([instructions](https://github.com/golang/vscode-go/blob/master/docs/troubleshooting.md#collect-gopls-information)) and post them in a new issue (if you're able to share them). Detailed gopls logs are best, so please add `-rpc.trace` to `go.languageServerFlags`.

For reference, this bug is referring to an error that appears like this:

![79361744-b837eb00-7f13-11ea-8bae-9683b4b61afc](https://user-images.githubusercontent.com/5856771/84923822-b6b2ac80-b095-11ea-85f4-9a98823fd7b5.png)

This issue consolidates https://github.com/microsoft/vscode-go/issues/3179, https://github.com/microsoft/vscode-go/issues/3063, https://github.com/microsoft/vscode-go/issues/3105, https://github.com/microsoft/vscode-go/issues/3141, https://github.com/microsoft/vscode-go/issues/3170, etc.

**Comment by mkeeler** [comment]
I am running into this issue and do not have any of those extensions installed.

VS Code:   v1.46.1
vscode-go: v0.14.4
gopls:     v0.4.1

The trace while this was happening is here: https://gist.github.com/mkeeler/fc6c5c3d910479a865c9b29dccc60a5b

There are several times where the getting the response from gopls is taking over 3 seconds and there are some RPCs being cancelled all the while.

For me the UI says `Saving '<various file names>': Getting code actions from "Go"`

**Comment by stamblerre** [comment]
Does this happen consistently / do you have a public repository in which this reproduces?

**Comment by mkeeler** [comment]
@stamblerre it happens every day multiple times throughout the day. I am almost always editing github.com/hashicorp/consul when it occurs. Not on every save though. So far it seems like the longer my editing session is the more I run into the problem.

**Comment by stamblerre** [comment]
Thanks for the report and repro case!

We recently made a fix in VS Code Go that we think will help with this. The next release is coming soon ([prerelease here](https://github.com/golang/vscode-go/releases/tag/v0.15.0-rc.2), [installation instructions here](https://github.com/golang/vscode-go/blob/v0.15.0-rc.2/docs/nightly.md#testing-pre-releases)). We also started ignoring `vendor` directories in the coming version of `gopls` - can you try out the pre-release? Download it by running `GO111MODULE=on go get golang.org/x/tools/gopls@v0.4.2-pre2`.

**Comment by mkeeler** [comment]
Just installed both. 🤞

**Comment by mkeeler** [comment]
There seems to be an issue with Go Nightly. After disabling the builtin one and enabling nightly most of the Go functionality is gone. No more symbols are available:

<img width="513" alt="Screen Shot 2020-06-26 at 10 05 52 AM" src="https://user-images.githubusercontent.com/1641976/85865929-a33dca80-b794-11ea-8e8e-c0335e2b08f8.png">

Other Go tasks fail to execute via the control panel: "Go: Test File" 

<img width="390" alt="image" src="https://user-images.githubusercontent.com/1641976/85865883-915c2780-b794-11ea-81b2-e479da2f1d3c.png">

Inline options to run individual tests are gone:

<img width="680" alt="Screen Shot 2020-06-26 at 10 06 37 AM" src="https://user-images.githubusercontent.com/1641976/85865985-bb154e80-b794-11ea-8666-0c5e93e52885.png">

I think the reason might be that its not starting gopls. I tried restarting via the command palette and got this error:

<img width="389" alt="Screen Shot 2020-06-26 at 10 07 54 AM" src="https://user-images.githubusercontent.com/1641976/85866110-e861fc80-b794-11ea-8f27-591bbc95181a.png">

**Comment by stamblerre** [comment]
Sorry about that - we had a bad release of the Nightly today so you must've caught it at the wrong time. I think the more important fixes are in `gopls` itself, so using the regular extension should be fine.

**Comment by mkeeler** [comment]
@stablerre With just the `gopls` prerelease things seemed fine for a while but recently I have started getting many multi-second saves. Nothing quite as bad as up to ~30 seconds before but now its more like 4-7s.

**Comment by stamblerre** [comment]
Thank you for the update! I believe the fixed version of the Nightly has been released, so if you could try that out too that would be helpful in diagnosing. Your repository doesn't have nested modules, does it?

**Comment by mkeeler** [comment]
It does have nested modules. If that is the cause here it would be one more reason to consider moving away from a multi-module repo.

**Comment by stamblerre** [comment]
We do typically recommend avoid multi-module repositories, but they shouldn't cause fundamental problems for `gopls`. Just to get a better understanding of your workflow, do you add a VS Code workspace folder for each module?

**Comment by mkeeler** [comment]
I typically just open the top level module in a workspace and edit everything from there. 

As for the nested modules, they exist in the same repo to make development easier (1 PR can implement an HTTP API and the API client additions at the same time). The Consul repo (along with the other HashiCorp product repos) includes an API client module. For our users we didn’t want them to need to pull in all of Consuls dependencies just to use our API, hence the sub module. We are on the verge of getting rid of the modules and having some automation at release time to push the directory from the repo over to a separate repo that can contain the module. I wish it wasn’t necessary but nested modules just causes too many headaches.

**Comment by mkeeler** [comment]
So I updated to Go Nightly v2020.6.2914 and if anything I am actually experiencing issues more often and more severely. I updated the gist here: https://gist.github.com/mkeeler/fc6c5c3d910479a865c9b29dccc60a5b.

**Comment by hyangah** [comment]
@mkeeler thanks for the update. that's disappointing - you mean Nightly shows popup on save more frequently with both versions of gopls?

With the old version of gopls - the very first `documentLink` request in the log took 34sec (probably blocked by previous requests). That indicates the `gopls` was broken way before it.

With the new version of gopls -  I noticed that the codeAction request failed with

```
 /Users/mkeeler/Code/repos/consul/auto-config/cert-generation/agent/consul/cluster_endpoint.go:149:28: missing ',' before newline in composite literal

[Error - 4:47:44 PM] Request textDocument/codeAction failed.
```

I still need to figure out what VS Code will do with the file save if the codeAction triggered on save action fails. But, if you fix the error in the `cluster_endpoint.go:149` before saving, do you still see the problem?

**Comment by mkeeler** [comment]
I didn't go back and test the v0.4.1 of gopls. Last week I updated from v0.4.1 to v0.4.2-pre2. That didn't fix the issue so first thing this morning I updated to using the Go Nightly extension instead of the non-Nightly one.

I will try to gather some more context about what is happening when it occurs to see if there is any correlation between compilation errors and it taking a long time.

**Comment by bobheadxi** [comment]
Spent all of yesterday using the Nightly extension + gopls v0.4.2-pre2 while working on https://github.com/sourcegraph/sourcegraph where I was previously encountering this problem and it seems like the problem is resolved for this specific case at least 😁 thanks for the update @stamblerre !

**Comment by mkeeler** [comment]
I wonder if it has something to do with the repo I am working on. Sometimes after a save I see gopls taking up to 5 full cores of CPU and consuming close to 7GB of ram.

<img width="563" alt="Screen Shot 2020-07-01 at 10 17 39 AM" src="https://user-images.githubusercontent.com/1641976/86254778-571fcb00-bb84-11ea-8f87-a11f79775e7e.png">

This is still with gopls v0.4.2-pre2

**Comment by stamblerre** [comment]
It's definitely possible. If you'd like us to take a look at the memory usage, you should see heap profiles in your temporary directory. Please upload them here, and we can take a look. /cc @heschik

**Comment by tylfin** [comment]
I just started experiencing slowness saving files with `gopls v0.4.2`

**Comment by stamblerre** [comment]
Thanks for reporting. https://github.com/golang/vscode-go/issues/299 has brought up issues with the fillstruct analyzer. Does disabling it help? You can add the following to your settings to disable it:

```
"gopls": {
    "analyses": {
        "fillstruct":false,
    }
}
```

**Comment by mkeeler** [comment]
I recently went back to Sublime Text using the same gopls version (v0.4.2) to see if it was experiencing the same problem. There are some difference in how the LSP is used between the two which make the user experience less jarring that I want to point out but before that I should mention that I  don't see any difference with regards to gopls resource utilization between how each is using it. With both CPU consumption tops out at 600% CPU and memory consumption at around 6 GB RSS. I think my issues largely stem from poor performance and responsiveness of gopls and not the vscode extension.

The main UX difference that makes the editing experience jarring in vscode is that the file isn't saved to disk until after all the gopls requests finish or are cancelled. If I save a file and switch over to a terminal to run some tests or build the code, I am often presented with the same compile errors I had just fixed and "saved". Once I save the file I expect its now on disk and ready for me to use outside of vscode but this isn't the case.

So my main request for vscode-go (and not gopls) would be to always save the file to disk first and run diagnostics after that.

**Comment by hyangah** [comment]
@mkeeler I somewhat agree except that many Go users expects some of the traditional features like formatting and goimports to run on save.
@stamblerre  Looking at my gopls trace, I see the code action was requested for organizeImports by default, with no diagnostics. Why does it trigger other diagnostics?

```
[Trace - 10:36:51.418 AM] Sending request 'textDocument/codeAction - (18)'.
Params: {"textDocument":{"uri":"file:///tools/internal/jsonrpc2/jsonrpc2.go"},"range":{"start":{"line":0,"character":0},"end":{"line":220,"character":0}},"context":{"diagnostics":[],"only":["source.organizeImports"]}}
```

**Comment by mkeeler** [comment]
@hyangah I still want format and import fixing on save. I was just thinking that for my usage it would be a better experience to do that async _after_ writing the bytes to disk rather than block disk writes waiting for it to finish. Although whether this is better or worse probably depends on your usage. I could see a scenario where you "save" and then go to commit the file to git and end up commiting a pre-formatted version. So maybe its just not good either way and the only real solution is to make gopls faster.

**Comment by Asgoret** [comment]
@stamblerre so...what I shall do for fixing it? I just rename package and save is broken. 
![Screenshot from 2020-07-09 13-27-27](https://user-images.githubusercontent.com/30697011/87028981-1d843b00-c1e8-11ea-9e96-0603abe842af.png)

**Comment by mkeeler** [comment]
FWIW, I have been using gopls v0.4.3 for the past 24 hours and haven't noticed the popups anymore.

**Comment by stamblerre** [comment]
> @stamblerre so...what I shall do for fixing it? I just rename package and save is broken.
> ![Screenshot from 2020-07-09 13-27-27](https://user-images.githubusercontent.com/30697011/87028981-1d843b00-c1e8-11ea-9e96-0603abe842af.png)
> 
> **SOLVED**: Just reboot PC and gopls re-read packages....It's little ridiculous...renaming packages crash golang server

@Asgoret: You shouldn't have to reboot your PC to reload `gopls` (you can use the `Go: Restart Language Server` command). Still, that issue doesn't seem related to this one. Please file a new issue if you see other issues with changing package names.

@mkeeler: Glad that things are working better with v0.4.3.

**Comment by Asgoret** [comment]
> > @stamblerre so...what I shall do for fixing it? I just rename package and save is broken.
> > ![Screenshot from 2020-07-09 13-27-27](https://user-images.githubusercontent.com/30697011/87028981-1d843b00-c1e8-11ea-9e96-0603abe842af.png)
> > **SOLVED**: Just reboot PC and gopls re-read packages....It's little ridiculous...renaming packages crash golang server
> 
> @Asgoret: You shouldn't have to reboot your PC to reload `gopls` (you can use the `Go: Restart Language Server` command). Still, that issue doesn't seem related to this one. Please file a new issue if you see other issues with changing package names.
> 
> @mkeeler: Glad that things are working better with v0.4.3.

Open issue here or in vscode repository?

**Comment by stamblerre** [comment]
> Open issue here or in vscode repository?

@Asgoret: In this repository.

**Comment by Wisgon** [comment]
I have a new discory. 
I also have the provlem of `Frequent pop-up on save: "Getting code actions: Go, Go, ..."`, I'm chinese, so in our country,google is outside our great fire wall... I can't open google.com unless I use a proxy.

I opened vscode normally, and the problem "Getting code actions: Go, Go, ..."  come out. Suddenly I find out that I'm not using proxy in vscode!!! So I turn on proxy, and, the miracle happen--the problem had gone, I can use gopls server fastly!!!

**Comment by tengge1** [comment]
I have a similar pop-up on save: 

```
The code in the workspace failed to compile (see the error message below).
If you believe this is a mistake, please file an issue: https://github.com/golang/go/issues/new.
go [-e -json -compiled=true -test=true -export=false -deps=true -find=false -- ./]: exit status 1: go: cannot find main module, but found .git/config in d:\Projects\GISCore
	to create a module there, run:
	go mod init
: packages.Load error
```

vscode-go version: 0.15.2

when I upgrade to 0.16.0, the problem no longer occurs. But intelligence does not work at all.

I wonder if there is a way to block this message.

**Comment by stamblerre** [comment]
@tengge1: Are you developing code in your GOPATH or in a Go module? It sounds like your workspace may not be correctly configured.

**Comment by tengge1** [comment]
@stamblerre The `go.mod` file is not in the root folder of the workspace.  It is in the `server/` folder. 
It is not a pure go project.

How can I config?

**Comment by stamblerre** [comment]
You will need to add a directory containing the `go.mod` file to your workspace. You can do this through View -> Add Folder to Workspace.

**Comment by tengge1** [comment]
@stamblerre Thank you very much. It helps a lot. I believe it is the best solution now. I'll try to change the project structure.

**Comment by vikram-rawat** [comment]
I can't save my main.go file at all. I get the error in a loop so fast that I can't even escape it. Any ctrl + S results in multiple pop ups and I can't work properly. This happened a month ago and hasn't been resolved even after the update of vscode and go 1.15 and gotools. Nothing worked.

```
Failed to save 'main.go': The content of the file is newer. Please compare your version with the file contents or overwrite the content of the file with your changes.
```

I just want to know how do I make it stop or has anyone reported this issue.

**Comment by stamblerre** [comment]
@vikram-rawat: I would suggest filing an issue with VS Code itself (https://github.com/microsoft/vscode). There are a few references online for help with this issue ([1](https://stackoverflow.com/questions/47892646/the-content-on-disk-is-newer-click-on-compare-to-compare-your-version-with-the), [2](https://code.visualstudio.com/docs/getstarted/tips-and-tricks#_preventing-dirty-writes)), but it's not caused by the Go extension.

**Comment by vikram-rawat** [comment]
I had an extentation of go imports that one i uninstalled. Now the errors are not so frequent but i still get an infinite saving go file.... 

And it never saves the file.

Vscode has become non usable since a few months. Please fix this issue anyhow

**Comment by hyangah** [comment]
@vikram-rawat Please follow the instruction https://github.com/golang/vscode-go/issues/236#issue-640579404 and share the settings and traces for investigation. Steps and code for repro will be very helpful. Thanks.

**Comment by vikram-rawat** [comment]
I deleted the Gopls I am better off writing without it? Thanks for letting me know that disabling extensions helps?

**Comment by upsampled** [comment]
I was having similar issues on Windows 10 Insider with Docker is WSL2 mode. Files were not saving and I was seeing the `Getting code action` notification. There was also a lot of hanging `setting up packages` when the container was first created. 

I seem to have gotten everything back to normal by switching Docker back to the legacy mode (turning off WSL2).

**Comment by segevfiner** [comment]
Seems like it is happening to me due to a long backlog of gopls reloading the packages with calls to `packages.Load` that are quite slow, especially when modifying the `go.mod` though other kinds of changes seem to trigger this as well. Until gopls works through all those reloads, it will not answer the request for code actions to auto format the file, which leads to this popup and the save hanging.

**Comment by hyangah** [comment]
@upsampled @segevfiner thank you for the updates. Both cases look like `gopls` were still busy loading packages and couldn't respond to the code action requests timely. 
`gopls` trace can help us investigating the issue. Do you mind sharing them with us?  `"go.languageServerFlags": ["-rpc.trace"]` in the setting will enable debug tracing.

**Comment by varungbt** [comment]
I too was struck with this issue. Ctrl+S (save) would take a lot of time or sometimes the editor got struck waiting forever. 

I had to reboot my machine and the issue went away.

**Comment by segevfiner** [comment]
It looks like this also happens when the parsing, type checking on typing of every single letter takes long enough that it starts to back up and you have to wait for all of it to complete. gopls doesn't debounce it, or tier the level or parsing it does at all. You can even see the error squiggles advancing letter by letter as long as you type at any reasonable speed.

Worse still are cases when it tries to reload the packages themselves. `packages.Load` is really slow, and when it does a repeated sequence of them on every letter typed... that's just slow.

I would have shared logs, but I'm working on a propriety code base, so will need to anonymize them somehow.

**Comment by hyangah** [comment]
@segevfiner @stamblerre  It sounds like the same issue as https://github.com/golang/go/issues/42932

**Comment by JabinGP** [comment]
> I have a new discovery.
> I also have the problem of `Frequent pop-up on save: "Getting code actions: Go, Go, ..."`, I'm Chinese, so in our country,google is outside our great fire wall... I can't open google.com unless I use a proxy.
> 
> Today I opened vscode, the problem "Getting code actions: Go, Go, ..." came out as usual. I closed vscode and reopen it, the problem still there... After a few times reopening vscode, suddenly I find out that I'm not using proxy in vscode!!! So I turn on proxy, and, the miracle happen--the problem had gone, I can use gopls server fastly!!! All the things be OK!
> 
> Then After a few minutes, I closed vscode, and waited for some minutes. I reopened it, everything still working fine!!!

Works for me, thank you！I have used the proxy on the router to slove this problem.

**Comment by AndrewRayCode** [comment]
I get this popup for almost every save, and it takes up to a minute to complete

> Saving 'file.go': Getting code actions from ''Go', 'GitHub Pull Requests and Issues'' (configure).

The configure link takes me to an empty part of my vscode setup.

Code introspection can also take a minute or more to load.

If I `pkill -f gopls` I'm able to save my file, and then the whole process starts over again.

Is there a way to debug the long file save issue?

**Comment by stamblerre** [comment]
@AndrewRayCode: Please see the comment at the top of this issue (https://github.com/golang/vscode-go/issues/236#issue-640579404) and follow the steps there. I would recommend opening a new issue.

**Comment by AndrewRayCode** [comment]
When I add this to my settings:

```
"go.buildFlags": ["-logfile", "-rpc.trace"],
```

My gopls server starts erroring with:

```
[Error - 2:23:02 PM] 2021/06/10 14:23:02 initial workspace load failed: err: exit status 2: stderr: flag provided but not defined: -logfile
usage: go list [-f format] [-json] [-m] [list flags] [build flags] [packages]
```

Same error for
```
"gopls": {"build.buildFlags": ["-logfile", "-rpc.trace"]},
```

The "documentation for your editor" doesn't have a clear place to put gopls flags, is that the right flag? I don't see any documentation on how to do the debugging steps.

**Comment by stamblerre** [comment]
@AndrewRayCode: Sorry for the old link at the top there, the setting you actually want is here: https://github.com/golang/vscode-go/blob/master/docs/troubleshooting.md#collect-gopls-information.

**Comment by nadgowdas** [comment]
Is there a solution/hack available for this problem ?

**Comment by stamblerre** [comment]
@nadgowdas: Please see the instructions in the first comment of the issue. If that doesn't resolve the problem, please open a new issue.

**Comment by stamblerre** [comment]
I think that this issue is no longer useful to keep open. Anyone who runs into this problem should open a new issue.
